### ISSUE TITLE
Bug: [@typescript-eslint/no-invalid-void-type] false positive on generics

### ISSUE THREAD

**Issue by G-Rath** [typescript-eslint/typescript-eslint#7227]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwFsAPAHgBUA%2BACgEoAueMgbgFgAod4kgNxy2Go1WHNhBwBzAIKpgAJRAZkMVDz4CqaUIiyoQwIe3YQF8XvwAKMHASwBnEA0vW7IVfwrwAvPF0B3eE429m7qtF4eAN7wAL4GoiZmwACyUAAODCmpJLYYMDriADSmah7efvCZ2bn5RYmCTEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KImjQO0SODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
declare function mx<T>(): T;

mx<void>();

logAndReturn<void>(undefined);

let voidPromise: Promise<void> = new Promise<void>(() => { });
let voidMap: Map<string, void> = new Map<string, void>();
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-invalid-void-type": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expected all of these to be considered valid uses of `void`

### Actual Result

Only the last two uses are considered valid.

### Additional Info

Maybe I'm missing something but I've read the docs a couple of times and the issues, and I can't find anything explaining why this general use of `void` is being flagged - or inversely, anything confirming that there is more logic than just "is this a generic?".

Importantly the `logAndReturn` is taken from the documentation itself for showcasing

> The following patterns are considered warnings with { allowInGenericTypeArguments: false }:

Which implies to me those patterns should _not_ be considered warnings if `allowInGenericTypeArguments` is true. This seems to be happening from v5 onwards.

**Comment by bradzacher** [comment]
This is working as expected - though the documentation is quite unclear on this point.

The option should probably be `allowInGeneticTypeArgumentsForTypes` because it specifically only applies to generic types as part of a type reference and does not apply to function calls and constructors.

I don't remember why it was built like this specifically. @JoshuaKGoldberg this was your baby - so you remember?

**Comment by JoshuaKGoldberg** [comment]
> `void is only valid as a return type or generic type argument. 3:4 - 3:8`

Amusingly, I think the message is mixing up _argument_ and _parameter_. Today the rule allows _parameters_. The complaint is on _arguments_.

> `allowInGenericTypeArgumentsForTypes`

Maybe we can:

1. v6.x: Add that option as proposed
2. v7.x: Breaking change the rule options:
    * Rename `allowInGenericTypeArguments` & `allowInGenericTypeParameters`
    * Allow them to be either catch-all booleans or granular objects, similar to `@typescript-eslint/prefer-nullish-coalescing`'s `ignorePrimitives` post-#7180?

> baby

...was this one of my rules? I don't remember getting too involved with it. Maybe more like a street cat I've been feeding? 😄

**Comment by G-Rath** [comment]
Sounds good to me!

**Comment by bradzacher** [comment]
@JoshuaKGoldberg sorry I thought you implemented this because you love to hate `void` 😄 

But no turns out it was you, @G-Rath!
And based on your original implementation - this is working as expected!
https://github.com/typescript-eslint/typescript-eslint/commit/f667ff1708d4ed28b7ea5beea742889da69a76d9#diff-0399d23c95309991615110bfd289120b920df533b4202452cb6ce10b6c89cc51R211-R220
https://github.com/typescript-eslint/typescript-eslint/blob/587ac306d9e53736ebe799f5b9edcb7dd030eed6/packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts#L197-L206

**Comment by G-Rath** [comment]
ah crap - I should have known, past-me is always out to get me...

As the original implementor though, does that mean I can arbitrarily change it? 😝

Seriously though I'll look over my original PR and try to figure out what I was thinking - I assume that I was bring over https://palantir.github.io/tslint/rules/invalid-void/ though I've not mentioned that in the PR body (but you @bradzacher mention "the original rule" in your initial review comment), and so that's probably where the logic came from rather than any strong personal opinions on my part that I've since unknowingly changed my mind about

**Comment by G-Rath** [comment]
I've had a bit of a further look over this, and I think actually it could be argued this is a general bug maybe combined with some bad/too-generic naming because `new Func<void>` is allowed for some reason (I assume for `new Promise<void>`).

If you're not happy to just change the logic to include functions without `new` as it being a bug, then I think probably it makes sense to add `allowInGenericTypeArgumentsForFunctions` - arguably the ideal would be that the existing option was actually `allowInGenericTypeArgumentsForTypes`, and then maybe `allowInGenericTypeArguments` just passes on to both options but that's probably not worth it.

What do folks think?

**Comment by binoche9** [comment]
Stumbling upon here because I also just ran into this while doing the `void` removal conquest in my own codebase 😄 

I'd be happy with any solution that makes this rule stop flagging these, but I'd add a vote for having these all controlled by a single `allowInGenerics` option (it seems to be how [TSLint](https://palantir.github.io/tslint/rules/invalid-void/) documented it at least), as the advantages/disadvantages of the different solutions seems pretty unclear to me.

From the example above, it's really confusing that

```ts
const x = mx<void>
```

works, but

```ts
const x = mx<void>()
```

doesn't

**Comment by JoshuaKGoldberg** [comment]
https://github.com/typescript-eslint/typescript-eslint/issues/7227#issuecomment-1654593588 makes sense to me. We can treat it as a breaking change for v7 to give ourselves a bit more flexibility in option naming. Thoughts @typescript-eslint/triage-team? 

> If there's a PR that allows void in all type parameters would it be accepted?

No, I think that'd be too much. Much of the time `void` isn't the right option (e.g. `Array<void>`).

**Comment by bradzacher** [comment]
> but i don't understand why anyone would ever have good reason to set that to false

```ts
declare function foo<T>(arg: T): string;

foo<void>(undefined);
```

An example of why the rule bans it.

A large number of generics are not specifically in a return type position in functions. There's a lot that are as well though. The isssue is that we can't tell purely based on syntax - the only way to "correctly" detect would be with type info and even then things can get dubious.

**Comment by 43081j** [comment]
That's true. I would have probably resorted to using type info, resolving function parameters' types and checking if they resolve to only void. 

If we add an option, do we want to default it to true? i.e allow void in all type parameters of call expressions?

**Comment by JoshuaKGoldberg** [comment]
Brad and I chatted briefly - we're slowly starting to sour on this rule as there are a couple issues (https://github.com/typescript-eslint/typescript-eslint/issues/6547, https://github.com/typescript-eslint/typescript-eslint/issues/7227) that are surprisingly intricate with it.

I also filed https://github.com/typescript-eslint/typescript-eslint/issues/8113 to track reworking the rule more holistically. I'm going to close & lock this issue for now so that we can focus on that discussion. Please do post there if you have more information to add. If it goes nowhere I'll re-open this one. Thanks all!

>  If we add an option, do we want to default it to true? i.e allow void in all type parameters of call expressions?

Great question. The proposal in #8113 is that it's always the behavior, and not an option. Let's discuss over there if there are reasons why we wouldn't want that to be the case.
